File: src/gepa/rpc/sdk/typescript/examples/omni.ts
/**
 * End-to-end example for the RunOptimizationOmni endpoint.
 *
 * Prereqs:
 *   1. gepa-rpc server running: `python -m gepa_rpc.cli --port 50051`
 *   2. OPENAI_API_KEY set in the server's environment.
 *
 * Run:
 *   npx tsx examples/omni.ts
 *
 * Task: optimize a system prompt that classifies a sentiment as positive or negative.
 * The evaluator checks whether the candidate prompt causes the model to output
 * the correct label — here simulated with a keyword match stand-in.
 */

import { Client, type Example } from "../src";

const dataset: Example[] = [
  { id: "1", fields: { text: "I love this product!", label: "positive" } },
  { id: "2", fields: { text: "This is terrible.", label: "negative" } },
  { id: "3", fields: { text: "Absolutely fantastic experience.", label: "positive" } },
  { id: "4", fields: { text: "Worst purchase I have ever made.", label: "negative" } },
];

async function main(): Promise<void> {
  const client = new Client({ target: "localhost:50051" });

  try {
    const result = await client.optimizeOmni({
      runId: `omni-demo-${Date.now()}`,
      seedCandidate: "Classify the sentiment of the following text as positive or negative.",
      dataset,
      objective: "Maximize accuracy of sentiment classification.",
      reflectionLm: "openai/gpt-4o-mini",
      maxEvals: 30,

      evaluate: async ({ candidate, batch }) => {
        const scores: number[] = [];
        const sideInfos: Record<string, unknown>[] = [];

        for (const ex of batch) {
          const text = ex.fields.text ?? "";
          const label = ex.fields.label ?? "";

          // Stand-in evaluator: score 1.0 if the candidate mentions both
          // "positive" and "negative" (i.e. it looks like a classification prompt),
          // and the label keyword appears in the candidate or text.
          const lowerCandidate = candidate.toLowerCase();
          const isClassificationPrompt =
            lowerCandidate.includes("positive") || lowerCandidate.includes("negative") ||
            lowerCandidate.includes("sentiment") || lowerCandidate.includes("classify");

          const score = isClassificationPrompt ? 1.0 : 0.0;
          scores.push(score);
          sideInfos.push({ text, label, score });
        }

        return { scores, sideInfos };
      },

      onProgress: (u) => {
        console.log(
          `[progress] ${u.evalsUsed}/${u.maxEvals} best=${u.bestScore.toFixed(3)} | "${u.bestCandidate.slice(0, 60)}..."`
        );
      },
    });

    console.log("\noptimization complete");
    console.log("  runId:        ", result.runId);
    console.log("  bestScore:    ", result.bestScore);
    console.log("  totalEvals:   ", result.totalEvals);
    console.log("  bestCandidate:", result.bestCandidate);
  } finally {
    client.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: src/gepa/rpc/sdk/rust/examples/omni.rs
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

use gepa_sdk::{Client, Example, OmniEvalResult, OmniOptimizeOpts};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let make_example = |id: &str, text: &str, label: &str| Example {
        id: id.to_string(),
        fields: HashMap::from([
            ("text".to_string(), text.to_string()),
            ("label".to_string(), label.to_string()),
        ]),
    };

    let dataset = vec![
        make_example("1", "I love this product!", "positive"),
        make_example("2", "This is terrible.", "negative"),
        make_example("3", "Absolutely fantastic experience.", "positive"),
        make_example("4", "Worst purchase I have ever made.", "negative"),
    ];

    let client = Client::new("localhost:50051");

    let millis = SystemTime::now().duration_since(UNIX_EPOCH)?.as_millis();

    let result = client
        .optimize_omni(OmniOptimizeOpts {
            run_id: format!("omni-demo-{millis}"),
            seed_candidate: Some(
                "Classify the sentiment of the following text as positive or negative."
                    .to_string(),
            ),
            dataset: Some(dataset),
            valset: None,
            objective: Some("Maximize accuracy of sentiment classification.".to_string()),
            reflection_lm: Some("openai/gpt-4o-mini".to_string()),
            engine: None, // defaults to "gepa"
            max_evals: 30,

            evaluate: |req: gepa_sdk::OmniEvalRequest| async move {
                let lower_candidate = req.candidate.to_lowercase();
                let is_classification_prompt = lower_candidate.contains("positive")
                    || lower_candidate.contains("negative")
                    || lower_candidate.contains("sentiment")
                    || lower_candidate.contains("classify");

                let scores = req
                    .batch
                    .iter()
                    .map(|_| if is_classification_prompt { 1.0 } else { 0.0 })
                    .collect();

                Ok(OmniEvalResult {
                    scores,
                    side_infos: None,
                })
            },

            on_progress: Some(Box::new(|u| {
                println!(
                    "[progress] {}/{} best={:.3} | \"{}...\"",
                    u.evals_used,
                    u.max_evals,
                    u.best_score,
                    &u.best_candidate.chars().take(60).collect::<String>()
                );
            })),
        })
        .await?;

    println!("\noptimization complete");
    println!("  runId:        {}", result.run_id);
    println!("  bestScore:    {}", result.best_score);
    println!("  totalEvals:   {}", result.total_evals);
    println!("  bestCandidate:{}", result.best_candidate);

    Ok(())
}
